### PR TITLE
Move more intrinsics to rustc_intrinsic

### PR DESCRIPTION
**Pull request by oli-obk**
cc https://github.com/rust-lang/rust/issues/63585

### CODE REVIEW

File: tests/ui/rfcs/rfc-2632-const-trait-impl/effects/minicore.rs [hunk 1]
```diff
@@ -398,17 +391,16 @@ impl<'a, T: ?Sized> Pin<&'a T> {
     }
 }
 
-
 impl<P: Deref> Pin<P> {
-    /* const */ fn as_ref(&self) -> Pin<&P::Target>
+    /* const */
+    fn as_ref(&self) -> Pin<&P::Target>
     where
-        P: /* ~const */ Deref,
```
Comment (compiler-errors) [inline tests/ui/rfcs/rfc-2632-const-trait-impl/effects/minicore.rs:405]: pls unformat this code, it deleted a bunch of comments

Comment (oli-obk) [reply tests/ui/rfcs/rfc-2632-const-trait-impl/effects/minicore.rs]: whoops

File: compiler/rustc_middle/src/hir/map/mod.rs [hunk 2]
```diff
@@ -1322,7 +1322,9 @@ impl<'hir> Visitor<'hir> for ItemCollector<'hir> {
 
     fn visit_item(&mut self, item: &'hir Item<'hir>) {
         if associated_body(Node::Item(item)).is_some() {
-            self.body_owners.push(item.owner_id.def_id);
+            if !self.tcx.has_attr(item.owner_id.def_id, sym::rustc_intrinsic_must_be_overridden) {
```
Comment (Nilstrieb) [inline compiler/rustc_middle/src/hir/map/mod.rs:1325]: Why was this necessary? I don't like it, can this be avoided by making the bodies of must_be_overriden intrinsics something else (like another intrinsic or intrinsics::unreachable^^)?

Comment (oli-obk) [reply compiler/rustc_middle/src/hir/map/mod.rs:1325]: Basically this avoids ever invoking typeck on those dummy bodies. We have various places that iterate over all body owners, some of them will try to invoke mir_const_qualifs, borrowck or just optimized_mir directly. I forget why we didn't need this check before this PR. const_eval_select was somehow special here...

Comment (Nilstrieb) [reply compiler/rustc_middle/src/hir/map/mod.rs:1325]: If we used `intrinsics::unreachable` (or maybe some short wrapper `must_override_body` function instead, would all of these problems go away? Hacking around not wanting to invoke typeck seems fragile

Comment (oli-obk) [reply compiler/rustc_middle/src/hir/map/mod.rs:1325]: Well, I don't want to return a value from `optimized_mir`, because that body could then end up in codegen somehow. So I added an assertion in mir_built to ensure we never build MIR for these intrinsics. That assert kept getting hit in the first commit, precisely because we assumed that any body owner will have a MIR body. Removing these intrinsics from MIR body owners sidesteps this issue in a fairly noninvasive way. It is fishy, I agree, but I also don't think it's worth adding body-less functions to HIR, AST and the parser, just for these intrinsics.

Comment (Nilstrieb) [reply compiler/rustc_middle/src/hir/map/mod.rs:1325]: Hm, how about making them just recurse? Typeck and MIR building etc should like that. It doesn't *guarantee* that the body won't be codegened, but I think that's okay.

Comment (oli-obk) [reply compiler/rustc_middle/src/hir/map/mod.rs:1325]: I just fear we'll screw up somewhere and end up actually invoking that body. An early impl of `rustc_intrinsic_must_be_overridden` had that issue for some cases. I guess I could codegen them all into an abort so it's obvious when it happens. It's very easy to debug with the assert in mir_built, but one can always add it again, and we can still avoid putting that MIR into metadata, causing ICEs in all crates but codegen.

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 3]
```diff
@@ -1701,8 +1701,10 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
         {
             for &local_def_id in tcx.mir_keys(()) {
                 if let DefKind::AssocFn | DefKind::Fn = tcx.def_kind(local_def_id) {
-                    record_array!(self.tables.deduced_param_attrs[local_def_id.to_def_id()] <-
-                        self.tcx.deduced_param_attrs(local_def_id.to_def_id()));
+                    if tcx.intrinsic(local_def_id).map_or(true, |i| !i.must_be_overridden) {
```
Comment (Nilstrieb) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:1704]: Does using a new `fake_body` or whatever intrinsic not help?

File: compiler/rustc_middle/src/hir/map/mod.rs (hunk 2 above)
Comment (Nilstrieb) [reply compiler/rustc_middle/src/hir/map/mod.rs]: This has probably come up before, but what about making them `extern "Rust"` with a `#[rustc_intrinsic]` attribute?

Comment (oli-obk) [reply compiler/rustc_middle/src/hir/map/mod.rs]: The reason we went away from foreign functions wasn't just the weird ABI, but also the problem that we can't specify safety or constness on foreign functions

File: compiler/rustc_metadata/src/rmeta/encoder.rs (hunk 3 above)
Comment (oli-obk) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:1704]: it does, but then I lose my nice assertion 😆

Comment (Nilstrieb) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:1704]: I feel like the assertion is less important than the maintenance cost of not sprinkling random checks like this in random places. If something does break and it chooses the runtime body it will crash immediately at runtime, which seems okay. But I'm not sure.

**Comment by oli-obk** [comment]
@rustbot author

Need to add the abort, otherwise screwing up and invoking the body causes UB

File: compiler/rustc_const_eval/src/interpret/terminator.rs [hunk 4]
```diff
@@ -212,6 +212,8 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 return Ok(());
             }
 
+            Abort => throw_machine_stop!(interpret::Abort),
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/terminator.rs:215]: I don't think this will work... Miri and const-eval assume that their own error type is the only type ever showing up here. `throw_machine_stop!` is not meant to ever be used in the core interpreter.

Can you make this a machine hook instead?

**Comment by Nilstrieb** [comment]
at that point I'd prefer to use the Unreachable terminator in this PR or add the Abort terminator in a separate PR (I'm inclined to use the Unreachable terminator since that should still often result in crashes and adding a new terminator seems like a lot of stuff for just the intrinsics, unless there are other use cases for it)

**Comment by oli-obk** [comment]
I removed the `Abort` terminator change
